lib/Gocdb_Services/User.php

```diff
@@ -422,31 +422,479 @@ public function deleteUser(\User $user, \User $currentUser = null) {
+        // Get list of tokens in order they are currently used
+        $myConfig1 = new \org\gocdb\security\authentication\MyConfig1();
+        $authTokenNames = $myConfig1->getAuthTokenClassList();
+
+        // Hardcoded authentication realms in same order as in token definitions
+        $x509Realms = ['X.509'];
+        if ($reducedRealms) {
+            $shibRealms = ['EGI Proxy IdP'];
+        } else {
+            $shibRealms = ['EUDAT_SSO_IDP', 'UK_ACCESS_FED', 'EGI Proxy IdP'];
+        }
+        $irisRealms = ['IRIS IAM - OIDC'];
+
+        // Add auth types to a list in the correct order
+        $authTypes = array();
+        foreach ($authTokenNames as $authTokenName) {
+            if (strpos($authTokenName, 'Shib') !== false) {
+                $authTypes = array_merge($authTypes, $shibRealms);
+            }
+            if (strpos($authTokenName, 'X509') !== false) {
+                $authTypes = array_merge($authTypes, $x509Realms);
+            }
+            if (strpos($authTokenName, 'IAM') !== false) {
+                $authTypes = array_merge($authTypes, $irisRealms);
+            }
+        }
+        return $authTypes;
+    }
+
+    /**
+     * Get one of the user's unique ID strings, favouring certain types
+     * If user does not have user identifiers, returns certificateDn
+     * @param \User $user User whose ID string we want
+     * @return string
+     */
+    public function getDefaultIdString($user) {
+
+        $authTypes = $this->getAuthTypes();
+        $idString = null;
+
+        // For each ordered auth type, check if an identifier matches
+        // Gets certifcateDn if no user identifiers and X.509 listed
+        foreach ($authTypes as $authType) {
+            $idString = $this->getIdStringByAuthType($user, $authType);
+            if ($idString !== null) {
+                break;
+            }
+        }
+
+        // If user only has unlisted identifiers, return first identifier
+        if ($idString === null) {
+            $idString = $user->getUserIdentifiers()[0]->getKeyValue();
+        }
+
+        return $idString;
+    }
+
+    /**
+     * Get a user's ID string of specified authentication type
+     * If user does not have user identifiers, returns certificateDn for X.509
+     * @param \User $user User whose ID string we want
+     * @param $authType authentication type of ID string we want
+     * @return string
+     */
+    public function getIdStringByAuthType($user, $authType) {
+
+        $identifiers = $user->getUserIdentifiers();
+        $idString = null;
+
+        // For each auth type, check if an identifier matches
+        foreach ($identifiers as $identifier) {
+            if ($identifier->getKeyName() === $authType) {
+                $idString = $identifier->getKeyValue();
+            }
+        }
+
+        // If no user identifiers and want X.509, return certificateDn
+        if (count($identifiers) === 0 && $authType === 'X.509') {
```

Comment:
Actually, I think this does preserve existing behavior.

Looking at where this method is called, for users who don't have any user identifiers yet, we **want** to return the the old certDN field - because it's the only identifier with have for them (and it is probably a certificate dn).